### PR TITLE
Check in FFI coverage list

### LINKED ISSUE #1945: Tracking issue: FFI completeness tool
This issue is tracking https://github.com/rust-diplomat/diplomat/issues/152

### PR DESCRIPTION
**Pull request by Manishearth**
Supersedes https://github.com/unicode-org/icu4x/pull/2403

trying to get this landed early

Fixes #1945

### CODE REVIEW

File: .github/workflows/build-test.yml [hunk 1]
```diff
@@ -270,6 +270,60 @@ jobs:
         crate: cargo-make
         version: latest
 
+    - name: Install Python
```
Comment (robertbastian) [inline .github/workflows/build-test.yml:273]: Should this run only after verify-ffi?

Comment (Manishearth) [reply .github/workflows/build-test.yml:273]: IMO no, i think this job is a lot like the fmt job: stuff is still likely to be working even if this fails, and it's easy to forget to run it and have it fail. It's good to have two statuses on each run, that's part of the goal behind splitting these.

**Review by robertbastian — APPROVED** [review]

**Comment by Manishearth** [comment]
It's unfortunate rustdoc-json needs nightly because it makes this task take longer than it needs to since it rebuilds all of ICU on nightly

File: .github/workflows/build-test.yml [hunk 2]
```diff
@@ -270,6 +270,63 @@ jobs:
...
+          ~/.cargo/bin/cargo-make.exe
+        key: ${{ runner.os }}-make-${{ steps.cargo-make-version.outputs.hash }}
+    - name: Install cargo-make
+      if: steps.cargo-make-cache.outputs.cache-hit != 'true'
+      uses: actions-rs/install@v0.1.2
+      with:
+        crate: cargo-make
+        version: latest
+
     - name: Install Python
```
Comment (sffc) [inline .github/workflows/build-test.yml:330]: Question: do you need Python for both jobs (ffi and verify-ffi)?

**Review by sffc — APPROVED** [review]

File: tools/scripts/ffi.toml [hunk 3]
```diff
@@ -27,6 +35,28 @@ description = "Verify that the minimum version of cargo-make exists"
 category = "ICU4X Development"
 install_crate = { crate_name = "cargo-make", binary = "cargo", test_arg = ["make", "--version"], min_version = "0.35.0" }
 
+[tasks.diplomat-coverage]
+description = "Produces the list of ICU APIs that are not exported through Diplomat"
+category = "ICU4X Development"
+script = '''
+cargo run --bin completeness --all-features > ffi/diplomat/tests/missing_apis.txt
```
[the remark this answers is not in the document]
Comment (sffc) [reply tools/scripts/ffi.toml:42]: Actually since this is a command that developers need to run themselves, it really should probably work on Windows

**Comment by Manishearth** [comment]
Ah, we can't use our pinned nightly because rustdoc-json is far more unstable

For now I'm using regular nightly. We may need to pin to a newer nightly instead.

**Comment by Manishearth** [comment]
We also ought to be updating our pinned nightly at some point anyway, but I'd rather not try that right now.

**Comment by sffc** [comment]
> Ah, we can't use our pinned nightly because rustdoc-json is far more unstable
> 
> For now I'm using regular nightly. We may need to pin to a newer nightly instead.

I would prefer not to have mixed nightlies in our repo, and definitely not have any cases where we have the latest nightly, since experience has shown that this causes flakiness, which is the most important thing to avoid.

Our currently nightly isn't _that_ old. Can we pin a slightly older version of rustdoc-json, or are there bleeding-edge features we need from it?

File: ffi/diplomat/ffi_coverage/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,38 @@
+# This file is part of ICU4X. For terms of use, please see the file
+# called LICENSE at the top level of the ICU4X source tree
+# (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+[package]
+name = "icu_ffi_coverage"
+description = "C interface to ICU4X"
+version = "1.0.0-alpha1"
```
Comment (sffc) [inline ffi/diplomat/ffi_coverage/Cargo.toml:8]: Nit: Don't mark this 1.0.0

**Review by sffc — APPROVED** [review]

**Review by sffc — APPROVED** [review]

**Pull request merged**